### PR TITLE
[OSDOCS-3996]: Port storage book to OSD and ROSA

### PR DESCRIPTION
**Pull request by bhardesty**
This PR ports the Storage book from OCP to the OSD and ROSA books using single-sourcing.

Version(s):
This applies to `main`, and `enterprise-4.12`.

Issue:
https://issues.redhat.com/browse/OSDOCS-3996

Links to docs previews:
- OCP: http://file.rdu.redhat.com/~bhardest/osdocs-3996/openshift-enterprise/storage/
- OSD: http://file.rdu.redhat.com/~bhardest/osdocs-3996/openshift-dedicated/storage/
- ROSA: http://file.rdu.redhat.com/~bhardest/osdocs-3996/openshift-rosa/storage/

QE review:
- [x] QE has approved this change.

### CODE REVIEW

**Comment by dustman9000** [comment]
@openshift/openshift-team-sd-sre-team-leads

File: storage/persistent_storage/persistent-storage-aws.adoc [hunk 1]
```diff
@@ -6,24 +6,50 @@ include::_attributes/common-attributes.adoc[]
 
 toc::[]
 
+ifdef::openshift-rosa,openshift-dedicated[]
+{product-title} (ROSA) clusters are prebuilt with four storage classes that use AWS Elastic Block Store (EBS) volumes. These storage classes are ready to use and some familiarity with Kubernetes and AWS is assumed.
```
Comment (yingzhanredhat) [inline storage/persistent_storage/persistent-storage-aws.adoc:10]: **(ROSA)** is not applicable to  OSD cluster

File: modules/dynamic-provisioning-change-default-class.adoc [hunk 2]
```diff
@@ -21,7 +23,7 @@ $ oc get storageclass
 [source,terminal]
 ----
 NAME                 TYPE
-gp2 (default)        kubernetes.io/aws-ebs <1>
+gp3 (default)        kubernetes.io/aws-ebs <1>
```
Comment (yingzhanredhat) [inline modules/dynamic-provisioning-change-default-class.adoc:26]: In 35 lines,the gp2 should be updated to gp3 too
`oc patch storageclass gp2 -p '{"metadata": {"annotations": {"storageclass.kubernetes.io/is-default-class": "false"}}}'`

Comment (duanwei33) [reply modules/dynamic-provisioning-change-default-class.adoc:26]: Here we just show how to update the default storageclass, we could even use common example like "Golden/Silvery" or "fast/slow", then no need to update in the further to map the real storageclass in OCP.
But we have to make sure the storageclass name specified in cmd below should be the same as example as mentioned by @yingzhanredhat

File: storage/persistent_storage/persistent-storage-aws.adoc [hunk 3]
```diff
@@ -6,24 +6,50 @@ include::_attributes/common-attributes.adoc[]
...
+
+Following are the four prebuilt storage classes:
```
Comment (duanwei33) [inline storage/persistent_storage/persistent-storage-aws.adoc:12]: I filed one jira(https://issues.redhat.com/browse/SDA-7296) for prebuilt storage classes, if that is accepted, doc may need to be updated as well.
cc @yingzhanredhat

File: storage/persistent_storage/persistent-storage-aws.adoc (hunk 1 above)
Comment (bhardesty) [reply storage/persistent_storage/persistent-storage-aws.adoc]: Fixed this.

File: modules/dynamic-provisioning-change-default-class.adoc (hunk 2 above)
Comment (bhardesty) [reply modules/dynamic-provisioning-change-default-class.adoc:26]: Done.

File: storage/persistent_storage/persistent-storage-aws.adoc (hunk 3 above)
Comment (bhardesty) [reply storage/persistent_storage/persistent-storage-aws.adoc:12]: Thanks for the heads-up!

File: modules/osd-persistent-storage-csi-efs-sts.adoc [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+NAME                                        READY   STATUS    RESTARTS   AGE
+cloud-credential-operator-78c9c575b-r6mmr   2/2     Running   0          6h33m
+----
+--
+
+.. Copy the `ccoctl` binary from the pod to a local directory.
++
+[source,terminal]
+----
+$ oc cp -c cloud-credential-operator openshift-cloud-credential-operator/<CCO-pod-name>:/usr/bin/ccoctl ./ccoctl
```
Comment (fahlmant) [inline modules/osd-persistent-storage-csi-efs-sts.adoc:46]: I'm not sure these commands would work for regular OSD/ROSA users

File: modules/osd-persistent-storage-csi-efs-sts.adoc [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+----
+
+.. To verify that `ccoctl` is ready to use, display the help file:
++
+[source,terminal]
+----
+$ ./ccoctl --help
+----
+
+. Create and save an EFS `CredentialsRequest` YAML file, such as shown in the following example:
```
Comment (fahlmant) [inline modules/osd-persistent-storage-csi-efs-sts.adoc:63]: I'm not sure this step would work for regular OSD/ROSA users

**Comment by bhardesty** [comment]
@kalexand-rh This is ready for peer review.

**Review by JoeAldinger — COMMENTED** [review]
Solid ifdef and ifndef work. A few small suggestions.

File: storage/persistent_storage/persistent-storage-aws.adoc [hunk 6]
```diff
@@ -6,24 +6,50 @@ include::_attributes/common-attributes.adoc[]
 
 toc::[]
 
+ifdef::openshift-rosa,openshift-dedicated[]
+{product-title} clusters are prebuilt with four storage classes that use AWS Elastic Block Store (EBS) volumes. These storage classes are ready to use and some familiarity with Kubernetes and AWS is assumed.
```
[the remark this answers is not in the document]
Comment (bhardesty) [reply storage/persistent_storage/persistent-storage-aws.adoc:10]: It's possible. Originally, there were two separate assemblies: one for Enterprise, and one for Dedicated. The Dedicated assembly had this paragraph and table, while the Enterprise assembly had a different paragraph. When I combined the two assemblies into one, I kept the two different paragraphs and differentiated them with ifdefs.

File: storage/persistent_storage/persistent-storage-aws.adoc [hunk 7]
```diff
@@ -6,24 +6,50 @@ include::_attributes/common-attributes.adoc[]
...
+
+| gp3-csi | ebs.csi.aws.com
+
+|===
+The gp3 storage class is set as default; however, you can select any of the storage classes as the default storage class.
+endif::openshift-rosa,openshift-dedicated[]
+
+ifndef::openshift-rosa,openshift-dedicated[]
 {product-title} supports AWS Elastic Block Store volumes (EBS). You can
 provision your {product-title} cluster with persistent storage by using link:https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/concepts.html[Amazon EC2].
```
[the remark this answers is not in the document]
Comment (bhardesty) [reply storage/persistent_storage/persistent-storage-aws.adoc:33]: I'm not positive what you mean by this - could you clarify? It does look like some of this content was originally written in the "one sentence per line" style. Is that what you're referring to?

File: _topic_maps/_topic_map_osd.yml [hunk 8]
```diff
@@ -195,14 +195,36 @@ Name: Storage
...
+  - Name: AWS Elastic Block Store CSI Driver Operator
+    File: persistent-storage-csi-ebs
+  - Name: AWS Elastic File Service CSI Driver Operator
+    File: osd-persistent-storage-aws-efs-csi
+  - Name: GCP PD CSI Driver Operator
+    File: persistent-storage-csi-gcp-pd
+  #- Name: GCP Filestore CSI Driver Operator
+  #  File: persistent-storage-csi-google-cloud-file
+- Name: Generic ephemeral volumes
+  File: generic-ephemeral-vols
```
[the remark this answers is not in the document]
Comment (bhardesty) [reply _topic_maps/_topic_map_osd.yml:225]: It wasn't.

**Comment by kalexand-rh** [comment]
@bhardesty, you need to squash.

**Comment by bhardesty** [comment]
@kalexand-rh Done.

**Comment by mburke5678** [comment]
@bhardesty Merge error.

**Comment by adellape** [comment]
@bhardesty Still needs rebase, removing `merge-review-needed` label for now; please re-add when it's ready to go. Thanks!

**Comment by mburke5678** [comment]
@bhardesty ^^

**Pull request merged**